sdk/tables/azure-data-tables/azure/data/tables/aio/_base_client_async.py

```diff
@@ -19,26 +27,189 @@
     CustomHookPolicy,
     NetworkTraceLoggingPolicy,
 )
-from azure.core.pipeline.transport import (
-    AsyncHttpTransport,
-    HttpRequest,
-)
 
 from ._authentication_async import _configure_credential
+from .._common_conversion import _is_cosmos_endpoint
+from .._constants import DEFAULT_STORAGE_ENDPOINT_SUFFIX
 from .._generated.aio import AzureTable
-from .._base_client import AccountHostsMixin, get_api_version, extract_batch_part_metadata
+from .._base_client import extract_batch_part_metadata, parse_query, format_query_string, get_api_version
 from .._error import (
     RequestTooLargeError,
     TableTransactionError,
     _decode_error,
     _validate_tablename_error,
 )
-from .._policies import StorageHosts, StorageHeadersPolicy
+from .._models import LocationMode
+from .._policies import CosmosPatchTransformPolicy, StorageHeadersPolicy, StorageHosts
 from .._sdk_moniker import SDK_MONIKER
 from ._policies_async import AsyncTablesRetryPolicy
 
 
-class AsyncTablesBaseClient(AccountHostsMixin):
+class AsyncAccountHostsMixin(object):  # pylint: disable=too-many-instance-attributes
+    def __init__(  # pylint: disable=too-many-statements
+        self,
+        account_url: str,
+        credential: Optional[Union[AzureNamedKeyCredential, AzureSasCredential, AsyncTokenCredential]] = None,
+        **kwargs: Any,
+    ) -> None:
+        try:
+            if not account_url.lower().startswith("http"):
+                account_url = "https://" + account_url
+        except AttributeError as exc:
+            raise ValueError("Account URL must be a string.") from exc
+        parsed_url = urlparse(account_url.rstrip("/"))
+        if not parsed_url.netloc:
+            raise ValueError(f"Invalid URL: {account_url}")
+
+        _, sas_token = parse_query(parsed_url.query)
+        if not sas_token and not credential:
+            raise ValueError("You need to provide either an AzureSasCredential or AzureNamedKeyCredential")
+        self._query_str, credential = format_query_string(sas_token, credential)
+        self._location_mode: str = kwargs.get("location_mode", LocationMode.PRIMARY)
+        self.scheme: str = parsed_url.scheme
+        self._cosmos_endpoint = _is_cosmos_endpoint(parsed_url)
+        self.account_name: Optional[str] = None
+        if ".core." in parsed_url.netloc or ".cosmos." in parsed_url.netloc:
```

Comment:
I'd like to reduce code duplication as much as possible. Are we able to move some of this out into a standalone function that could be shared between sync and async?

Reply:
I extracted the logic to get account and account name to a separate function in `_common_conversion.py`.